services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java

```diff
@@ -4596,38 +4596,80 @@ private void setStoreCurrentVersion(
     });
   }
 
+  boolean isCurrentRegionPartOfRegionFilter(String regionFilter) {
+    if (StringUtils.isEmpty(regionFilter)) {
+      return true;
+    }
+    String currRegionName = multiClusterConfigs.getRegionName();
+    Set<String> regionsFilter = parseRegionsFilterList(regionFilter);
+    return regionsFilter.contains(currRegionName);
+  }
+
   @Override
   public void rollForwardToFutureVersion(String clusterName, String storeName, String regionFilter) {
-    if (!StringUtils.isEmpty(regionFilter)) {
-      Set<String> regionsFilter = parseRegionsFilterList(regionFilter);
-      if (!regionsFilter.contains(multiClusterConfigs.getRegionName())) {
-        LOGGER.info(
-            "rollForwardToFutureVersion command will be skipped for store: {} in cluster: {}, because the region filter is {}"
-                + " which doesn't include the current region: {}",
-            storeName,
-            clusterName,
-            regionsFilter,
-            multiClusterConfigs.getRegionName());
-        return;
-      }
+    if (!isCurrentRegionPartOfRegionFilter(regionFilter)) {
+      LOGGER.info(
+          "Rolling forward will be skipped for store: {} in cluster: {}, as the region filter is {}",
+          storeName,
+          clusterName,
+          regionFilter);
+      return;
     }
     int futureVersion = getOnlineFutureVersion(clusterName, storeName);
     if (futureVersion == NON_EXISTING_VERSION) {
+      LOGGER.info(
+          "Rolling forward will be skipped for store: {} in cluster: {}, as there is no future version",
+          storeName,
+          clusterName);
       return;
     }
     storeMetadataUpdate(clusterName, storeName, store -> {
       if (!store.isEnableWrites()) {
         throw new VeniceException(
             "Unable to update store:" + storeName + " current version since store does not enable writes");
       }
+      // check whether the future version has enough ready-to-serve instances for all partitions in CV
       int previousVersion = store.getCurrentVersion();
+      Version futureVersionObj = store.getVersion(futureVersion);
+      int partitionCount = futureVersionObj.getPartitionCount();
+      int minActiveReplicas = futureVersionObj.getMinActiveReplicas();
+      String currentVersionKafka = Version.composeKafkaTopic(storeName, futureVersion);
+      HelixCustomizedViewOfflinePushRepository customizedViewRepository =
+          getHelixVeniceClusterResources(clusterName).getCustomizedViewRepository();
+      List<Integer> unFinishedPartitions = new ArrayList<>();
+      for (int partition = 0; partition < partitionCount; partition++) {
+        List<Instance> readyToServeInstances;
+        try {
+          readyToServeInstances = customizedViewRepository.getReadyToServeInstances(currentVersionKafka, partition);
+        } catch (Exception e) {
+          readyToServeInstances = Collections.emptyList();
```

Comment:
should we break here? add a long?

Reply:
Break yeah, but do we want to show all the partitions or probably the number of partitions not completed yet? If not, break can be added.